core/broadcast.go

```diff
@@ -0,0 +1,201 @@
+	hashes set
+	dealCh chan dkg.DealBundle
+	respCh chan dkg.ResponseBundle
+	justCh chan dkg.JustificationBundle
+	verif  verifier
+}
+
+type packet = dkg.Packet
+
+var _ dkg.Board = (*broadcast)(nil)
+
+// verifier is a type for  a function that can verify the validity of a dkg
+// Packet, namely that the signature is correct.
+type verifier func(packet) error
+
+func newBroadcast(l log.Logger, c net.ProtocolClient, own string, to []*key.Node, v verifier) *broadcast {
+	return &broadcast{
+		l:      l,
+		own:    own,
+		client: c,
+		to:     to,
+		dealCh: make(chan dkg.DealBundle, len(to)),
+		respCh: make(chan dkg.ResponseBundle, len(to)),
+		justCh: make(chan dkg.JustificationBundle, len(to)),
+		hashes: new(arraySet),
+		verif:  v,
+	}
+}
+
+func (b *broadcast) PushDeals(bundle *dkg.DealBundle) {
+	b.dealCh <- *bundle
+	h := hash(bundle.Hash())
+	b.l.Debug("broadcast", "push", "deal")
+	b.sendout(h, bundle)
+}
+
+func (b *broadcast) PushResponses(bundle *dkg.ResponseBundle) {
+	b.respCh <- *bundle
+	h := hash(bundle.Hash())
+	b.l.Debug("broadcast", "push", "response", bundle.String())
+	b.sendout(h, bundle)
+}
+
+func (b *broadcast) PushJustifications(bundle *dkg.JustificationBundle) {
+	b.justCh <- *bundle
+	h := hash(bundle.Hash())
+	b.l.Debug("broadcast", "push", "justification")
+	b.sendout(h, bundle)
+}
+
+func (b *broadcast) BroadcastDKG(c context.Context, p *drand.DKGPacket) (*drand.Empty, error) {
+	b.Lock()
+	defer b.Unlock()
+	addr := net.RemoteAddress(c)
+	dkgPacket, err := protoToDKGPacket(p.GetDkg())
+	if err != nil {
+		b.l.Debug("broadcast", "received invalid packet", "from", addr, "err", err)
+		return nil, errors.New("invalid packet")
+	}
+
+	hash := hash(dkgPacket.Hash())
+	if b.hashes.exists(hash) {
+		// if we already seen this one, no need to verify even because that
+		// means we already broadcasted it
+		b.l.Debug("broadcast", "ignoring duplicate packet", "from", addr, "type", fmt.Sprintf("%T", dkgPacket))
+		return new(drand.Empty), nil
+	}
+	if err := b.verif(dkgPacket); err != nil {
+		b.l.Debug("broadcast", "received invalid signature", "from", addr)
+		return nil, errors.New("invalid packet")
+	}
+
+	b.l.Debug("broadcast", "received new packet to broadcast", "from", addr, "type", fmt.Sprintf("%T", dkgPacket))
+	b.sendout(hash, dkgPacket)
+	b.dispatch(dkgPacket)
+	return new(drand.Empty), nil
+}
+
+func (b *broadcast) dispatch(p packet) {
+	switch pp := p.(type) {
+	case *dkg.DealBundle:
+		b.dealCh <- *pp
+	case *dkg.ResponseBundle:
+		b.respCh <- *pp
+	case *dkg.JustificationBundle:
+		b.justCh <- *pp
+	}
+}
+
+func (b *broadcast) sendout(h []byte, p packet) {
+	dkgproto, err := dkgPacketToProto(p)
+	if err != nil {
+		b.l.Error("broadcast", "can't send packet", "err", err)
+		return
+	}
+	// we register we saw that packet and we broadcast it
+	b.hashes.put(h)
+	proto := &drand.DKGPacket{
+		Dkg: dkgproto,
+	}
+	for _, n := range b.to {
+		if n.Address() == b.own {
+			continue
+		}
+		go func(nn *key.Node) {
```

Comment:
rather than a goroutine each message to a peer, consider channels for each peer. It'll help make sure that for individual peers messages stay in order, and limits growth of memory / goroutines.